src/VisualStudio/Core/Test.Next/Mocks/InProcRemoteHostClientFactory.cs

```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Remote;
+using Roslyn.Test.Utilities.Remote;
+
+namespace Roslyn.VisualStudio.Next.UnitTests.Mocks
+{
+    [ExportWorkspaceService(typeof(IRemoteHostClientFactory), layer: ServiceLayer.Host), Shared]
+    internal class InProcRemoteHostClientFactory : IRemoteHostClientFactory
+    {
+        public Task<RemoteHostClient> CreateAsync(Workspace workspace, CancellationToken cancellationToken)
+        {
+            return InProcRemoteHostClient.CreateAsync(workspace, cancellationToken);
```

Comment:
RemoteHostClientService is the one who cache. this is factory that create new one when asked. caller is the one who decide whether it will reuse last one or create new one.